File: app/src/main/java/com/alexkn/syntact/presentation/common/ListItemAdapter.java
package com.alexkn.syntact.presentation.common;

import com.alexkn.syntact.data.common.Identifiable;

import java.util.List;

import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

public abstract class ListItemAdapter<T extends Identifiable, S extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<S> {

    private final AsyncListDiffer<T> differ = new AsyncListDiffer<>(this, onDiff());

    private DiffUtil.ItemCallback<T> onDiff(){
        return new DiffCallback<>();
    }

    public void submitList(List<T> data) {
        differ.submitList(data);
    }

    @Override
    public int getItemCount() {
        return differ.getCurrentList().size();
    }

    public List<T> getList() {
        return differ.getCurrentList();
    }
}

File: app/src/main/java/com/alexkn/syntact/presentation/common/ListItemAdapter.kt
package com.alexkn.syntact.presentation.common

import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.alexkn.syntact.core.model.Identifiable

abstract class ListItemAdapter<T : Identifiable<*>, S : RecyclerView.ViewHolder> : RecyclerView.Adapter<S>() {

    private val differ = AsyncListDiffer(this, onDiff())

    val list: List<T>
        get() = differ.currentList

    private fun onDiff(): DiffUtil.ItemCallback<T> = DiffCallback()

    fun submitList(data: List<T>) = differ.submitList(data)

    override fun getItemCount(): Int = differ.currentList.size
}
